packages/hardhat-utils/src/internal/request.ts

```diff
@@ -0,0 +1,129 @@
+  const { dir, ext, name } = path.parse(filePath);
+
+  await mkdir(dir);
+
+  return path.format({
+    dir,
+    ext,
+    name: `tmp-${name}`,
+  });
+}
+
+export async function getBaseRequestOptions(
+  requestUrl: string,
+  { extraHeaders, abortSignal, queryParams }: RequestOptions = {},
+  dispatcherOrDispatcherOptions?: UndiciT.Dispatcher | DispatcherOptions,
+) {
+  const { Dispatcher } = await import("undici");
+  const dispatcher =
+    dispatcherOrDispatcherOptions instanceof Dispatcher
+      ? dispatcherOrDispatcherOptions
+      : await getDispatcher(requestUrl, dispatcherOrDispatcherOptions);
+
+  // We could use the global dispatcher if neither dispatcher nor dispatcherOptions were passed,
+  // but there's no way to configure it, so we don't do it.
+  // https://github.com/nodejs/undici/blob/961b76ad7cac17d23580d172702e11a080974f5d/lib/global.js#L9
+  return {
+    dispatcher,
+    headers: getHeaders(requestUrl, extraHeaders),
+    throwOnError: true,
+    ...(abortSignal !== undefined ? { signal: abortSignal } : {}),
+    ...(queryParams !== undefined ? { query: queryParams } : {}),
+  };
+}
+
+export function getHeaders(
+  requestUrl: string,
+  extraHeaders: Record<string, string> = {},
+) {
+  const headers: Record<string, string> = {
+    ...extraHeaders,
+    "User-Agent": extraHeaders["User-Agent"] ?? DEFAULT_USER_AGENT,
+  };
+
+  const authHeader = getAuthHeader(requestUrl);
+  if (authHeader !== undefined) {
+    headers.Authorization = authHeader;
+  }
+
+  return headers;
+}
+
+export function getAuthHeader(requestUrl: string) {
+  const parsedUrl = new URL(requestUrl);
+  if (parsedUrl.username === "") {
+    return undefined;
+  }
+
+  return `Basic ${Buffer.from(
+    `${parsedUrl.username}:${parsedUrl.password}`,
+  ).toString("base64")}`;
+}
+
+export async function getProxyDispatcher(
+  proxy: string,
+  options: Omit<UndiciT.ProxyAgent.Options, "uri">,
+): Promise<UndiciT.ProxyAgent> {
+  const { ProxyAgent } = await import("undici");
+
+  return new ProxyAgent({
+    uri: proxy,
+    ...options,
+  });
+}
+
+export async function getPoolDispatcher(
+  requestUrl: string,
+  options: UndiciT.Pool.Options,
+): Promise<UndiciT.Pool> {
+  const { Pool } = await import("undici");
+
+  const parsedUrl = new URL(requestUrl);
+  return new Pool(parsedUrl.origin, options);
+}
+
+export async function getBasicDispatcher(
+  options: UndiciT.Agent.Options,
+): Promise<UndiciT.Agent> {
+  const { Agent } = await import("undici");
+
+  return new Agent(options);
+}
+
+export function getBaseDispatcherOptions(
+  timeout: number = DEFAULT_TIMEOUT_IN_MILLISECONDS,
+  isTestDispatcher: boolean = false,
+): UndiciT.Client.Options {
+  // These have good defaults for production, but need to be tweaked to avoid hanging tests.
+  // https://github.com/nodejs/undici/blob/961b76ad7cac17d23580d172702e11a080974f5d/docs/docs/best-practices/writing-tests.md
+  const keepAliveTimeouts = isTestDispatcher
```

Comment:
I'm not sure about this approach. I think we should use a mock dispatcher in our tests, similar to what I've done in `packages/hardhat-utils/test/request.ts`. Maybe we could add a `getMockDispatcher` function in `packages/hardhat-utils/src/request.ts` to make sure the `MockAgent` is initialized with this options. @alcuadrado, wdyt?